docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -262,8 +262,166 @@ Every inter-canister call is guaranteed to receive a response, either from the c
 
+### Securely handle traps in callbacks
+
+#### Security Concern
+
+Traps / panics roll back the canister state, as described in Property 5 above. So any state change followed by a trap or panic can be risky. This is an important concern when inter-canister calls are made. If a trap occurs after an await to an inter-canister call, then the state is reverted to the snapshot before the inter-canister call’s callback invocation, and not to the state before the entire call.
+
+More precisely, suppose some state changes are applied and then an inter-canister call is issued. Also, assume that these state changes leave the canister in an inconsistent state, and that state is only made consistent again in the callback. Now if there is a trap in the callback, this leaves the canister in an inconsistent state.
+
+Here are two example security issues that can arise because of this:
+
+- Assume an inter-canister call is issued to transfer funds. In the callback, the canister accounts for having made that transfer by updating the balances in the canister storage. However, suppose the callback also updates some usage statistics data, which eventually leads to a trap when some data structure becomes full. As soon as that is the case, the canister ends up in an inconsistent state because the state changes in the callback are no longer applied and thus the transfers are not correctly accounted for.
+  ![example_highlighted_code](./_attachments/example_trap_after_await.png)
+  This example is also discussed in this [community conversation](https://www.youtube.com/watch?v=PneRzDmf_Xw&list=PLuhDt1vhGcrez-f3I0_hvbwGZHZzkZ7Ng&index=2&t=4s). 
+
+- Suppose part of the canister state is locked before an inter-canister call and released in the callback. Then the lock may never be released if the callback traps.
+Note that in canisters implemented in Rust with Rust CDK version 0.5.1, any local variables still go out of scope if a callback traps. The CDK actually calls into the ic0.call_on_cleanup API to release these resources. This helps to prevent issues with locks not being released, as it is possible to use Rust's Drop implementation to release locked resources, as we discuss in [Be aware that there is no reliable message ordering](https://internetcomputer.org/docs/current/developer-docs/security/rust-canister-development-security-best-practices#be-aware-that-there-is-no-reliable-message-ordering).
+
+#### Recommendation
+
+Recall that the responses to inter-canister calls are processed in the corresponding callback. If the callback traps, the cleanup (ic0.call_on_cleanup) is executed. When making an inter-canister call, the ICP reserves sufficiently many cycles to execute the response callback or cleanup (up to the instruction limit). A fixed fraction of the reservation is for the cleanup. Thus, a response or cleanup execution can never “run out of cycles”, but they can run into the instruction limit and trap.
```

Comment:
Do you actually know which percentage is reserved? This sounds like something we should document, maybe [here](https://internetcomputer.org/docs/current/developer-docs/gas-cost#canister-calls)?
Can you still hit that percentage limit in the `call_on_cleanup`?

Reply:
IIRC @robin-kunzler made these changes, with the goal of making the rest of the doc consistent with the addition of the journaling section. I think it's somewhat of a scope creep to improve this section in this PR and I suggest putting this in a different ticket.

Reply:
I haven't found this info anywhere in the IC docs. If we don't add it we should definitely try to have it added somewhere as it is important for people that try to use this cleanup to comply with the best practice.

This reservation also wasn't obvious to me from the IC code so we might just ask @ulan

Reply:
Add issue #2817. I recommend resolving this outside this PR.